File: Lab 5/Python/Task_1_Gaussian.py
import numpy as np
from scipy.integrate import quad
import matplotlib.pyplot as plt
import math
from sympy import *


def gauss_integral(mu, sigma):
    PDF = lambda x: np.exp(-(x - mu)**2/(2*sigma**2))/(sigma*math.sqrt(2*pi))
    total_prob, error_prob = quad(lambda x: np.exp(-(x - mu)**2/(2*sigma**2))/(sigma*math.sqrt(2*pi)), -np.inf, np.inf)
    mean, error_m = quad(lambda x: x*np.exp(-(x - mu)**2/(2*sigma**2))/(sigma*math.sqrt(2*pi)), -np.inf, np.inf)
    expectation_xsq_fx, error_var = quad(lambda x: x**2*np.exp(-(x - mu)**2/(2*sigma**2))/(sigma*math.sqrt(2*pi)), -np.inf, np.inf)

    variance = expectation_xsq_fx - mean**2
    return [PDF, total_prob, mean, variance]


mu = 1
sigma = 1
xmin = -4
xmax = 6
fs = 10

x_plot = np.arange(xmin, xmax, 1/fs)
PDF, total_probability, mean, variance = gauss_integral(mu, sigma)
print('Total Probability of the PDF', total_probability,
      '\nMean of the PDF', mean,
      '\nVariance of the PDF', variance)

plt.figure(1)
plt.plot(x_plot, PDF(x_plot))
plt.title('Gaussian PDF')
plt.xlabel('Running Variable')
plt.ylabel('Amplitude')
plt.show()

# Additional -- Finding PDF using symbolic integration
x = Symbol('x')
PDF = exp(-(x - mu)**2/(2*sigma**2))/(sigma*sqrt(2*pi))
CDF = integrate(PDF, x)
print('The CDF is', CDF)

File: Lab 5/MATLAB/Task_1_Gaussian.m
% PDF of Normal Gaussian RV

sigma = 1;
mu = 0; 
fs = 10;
xmin = -5;
xmax = 5;
xq = 1;

% Calculating probability, mean and variance using the integral formulae
% for CDF and expectation
[total_prob,mean_val,var_val] = GaussRV(mu,sigma,xmin,xmax,fs);

disp('The total probability (should be 1):')
disp(total_prob)
disp('Mean for the distribution:')
disp(mean_val)
disp('Variance for the distribution:')
disp(var_val)

function [q_0, mean_val,var_val] = GaussRV(mu,sigma,xmin,xmax,fs)
    x = xmin:1/fs:xmax;
    pdf = exp(-(x - mu).^2/(2*(sigma)^2))/(sigma*sqrt(2*pi));
    pdf_fn = @(t)exp(-(t - mu).^2/(2*(sigma)^2))/(sigma*sqrt(2*pi));
    exp_x_fn = @(t)t.*exp(-(t - mu).^2/(2*(sigma)^2))/(sigma*sqrt(2*pi));
    exp_xsq_fn = @(t)t.^2.*exp(-(t - mu).^2/(2*(sigma)^2))/(sigma*sqrt(2*pi));

    figure()
    plot(x,pdf)

    title("Gaussian RV")
    xlabel("Running variable")
    ylabel("PDF")
    grid on

    q_0 = integral(pdf_fn,-inf,inf);
    mean_val = integral(exp_x_fn,-inf,inf);
    var_val = integral(exp_xsq_fn,-inf,inf) - mean_val^2;
end